shared/actions/signup.js

```diff
@@ -1,366 +1,304 @@
 // @flow
 import logger from '../logger'
+import * as Constants from '../constants/signup'
 import * as LoginGen from './login-gen'
 import * as SignupGen from './signup-gen'
+import * as Saga from '../util/saga'
 import * as RPCTypes from '../constants/types/rpc-gen'
 import HiddenString from '../util/hidden-string'
 import {trim} from 'lodash-es'
 import {isMobile} from '../constants/platform'
 import {isValidEmail, isValidName, isValidUsername} from '../util/simple-validators'
 import {loginTab} from '../constants/tabs'
 import {navigateAppend, navigateTo} from '../actions/route-tree'
+import type {RPCError} from '../engine/types'
+import type {TypedState} from '../constants/reducer'
 
-function nextPhase() {
-  return (dispatch, getState) => {
-    // TODO careful here since this will not be sync on a remote component!
-    const phase: string = getState().signup.phase
-    dispatch(navigateAppend([phase], [loginTab, 'signup']))
+const checkInviteCode = (action: SignupGen.CheckInviteCodePayload) =>
+  Saga.call(
+    RPCTypes.signupCheckInvitationCodeRpcPromise,
+    {invitationCode: action.payload.inviteCode},
+    Constants.waitingKey
+  )
+const checkInviteCodeSuccess = (_, action: SignupGen.CheckInviteCodePayload) =>
+  Saga.sequentially([
+    Saga.put(SignupGen.createCheckInviteCodeDone({inviteCode: action.payload.inviteCode})),
+    Saga.put(navigateTo([loginTab, 'signup', 'usernameAndEmail'])),
+  ])
+const checkInviteCodeError = () =>
+  Saga.put(SignupGen.createCheckInviteCodeDoneError({errorText: "Sorry, that's not a valid invite code."}))
+
+const requestAutoInvite = () =>
+  Saga.call(RPCTypes.signupGetInvitationCodeRpcPromise, undefined, Constants.waitingKey)
+const requestAutoInviteSuccess = (inviteCode: string) =>
+  Saga.put(SignupGen.createCheckInviteCode({inviteCode}))
+const requestAutoInviteError = () => Saga.put(navigateTo([loginTab, 'signup', 'inviteCode']))
+
+const requestInvite = (action: SignupGen.RequestInvitePayload) => {
+  const {email, name} = action.payload
+  const emailError = isValidEmail(email)
+  const nameError = isValidName(name)
+  if (emailError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError, name, nameError: ''}))
+  }
+  if (nameError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError: '', name, nameError}))
   }
+
+  return Saga.call(
+    RPCTypes.signupInviteRequestRpcPromise,
+    {email: email, fullname: name, notes: 'Requested through GUI app'},
+    Constants.waitingKey
+  )
 }
 
-function startRequestInvite() {
-  return (dispatch: Dispatch) => {
-    dispatch(SignupGen.createStartRequestInvite())
-    dispatch(nextPhase())
+const requestInviteSuccess = (
+  result: SignupGen.RequestInviteDonePayload | void,
+  action: SignupGen.RequestInvitePayload
+) => {
+  // rpc returns undefined, dispatches above on error return the type
```

Comment:
I don't understand this comment (and similar below), or why a non-null `result` means do nothing (and similar below)

Reply:
requestInvite returns either a Saga.put or the results of the Saga.call on success. Therefore results here (as typed) is `SignupGen.RequestInviteDonePayload | void`.  originally i was throwing an error in that case but the typing was worse and it was more code but i could put it back that way

Reply:
i'll go back to the throwing way since i agree this is more confusing

Reply:
actually i have a larger change that i think fixes the fundamental issue i'll discuss in part 5